## Predict using correlation coefficient

### Given x and y and linear regression equation y = b*x + a + e , solve for a and b such that sum(e ^ 2) is minimum

In [1]:
import os
import pandas as pd
import numpy as np

In [2]:
os.chdir("C:/Users/satish/Downloads")

cars = pd.read_csv("cars.csv")

In [3]:
x= cars["Weight"]
y= cars["MPG"]

sum_yi = np.sum(y)
sum_xi = np.sum(x)
sum_xiyi = np.sum(x*y)
sum_xi_squared = np.sum(x**2)
n = len(x)

In [4]:
a  = ((sum_yi * sum_xi_squared) - (sum_xi * sum_xiyi))/(n * sum_xi_squared - (sum_xi**2))
print(a)

46.325305341697934


In [5]:
b = (n * sum_xiyi - (sum_xi * sum_yi))/(n * sum_xi_squared - (sum_xi**2))
print(b)

-0.00781162853863321


### Thus we've found the coefficient a and intercept b using OLS method. For one unit increase in x, there'll be 0.007 units decrease in y. If x is 0, y will be 46.325

## Predict Yi

In [6]:
#Predicting the target value
yi = a + (b* x)

In [7]:
# Display the predicted and actual value
result = {"regression MPG":yi, "actual MPG" : y}
result = pd.DataFrame(result)
result

regression MPG  actual MPG
0         18.953359        18.0
1         17.476961        15.0
2         19.484550        18.0
3         19.507985        16.0
4         19.382999        17.0
..              ...         ...
401       24.530862        27.0
402       29.686537        44.0
403       28.397618        32.0
404       25.819780        28.0
405       25.077676        31.0

[406 rows x 2 columns]

In [12]:
from sklearn import metrics
import math

mse = metrics.mean_squared_error(y, yi)
rmse = math.sqrt(mse)
print("RMSE: ", rmse)

print("Target range: ", y.min()," to ", y.max())

RMSE:  5.171650947982947
Target range:  0.0  to  46.6


### Root mean squared error of model is 5.17, which is good, since range is 0 to 46.6